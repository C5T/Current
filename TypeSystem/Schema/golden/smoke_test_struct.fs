// Usage: `fsharpi -r Newtonsoft.Json.dll schema.fs`
// (*
// open Newtonsoft.Json
// let inline JSON o = JsonConvert.SerializeObject(o)
// let inline ParseJSON (s : string) : 'T = JsonConvert.DeserializeObject<'T>(s)
// *)

type Primitives = {
  a : byte
  b : uint16
  c : uint32
  d : uint64
  e : sbyte
  f : int16
  g : int32
  h : int64
  i : char
  j : string
  k : float
  l : double
  m : bool
  n : int64  // microseconds.
  o : int64  // milliseconds.
}

type A = {
  a : int32
}

type B = {
  a : int32
  b : int32
}

type X = {
  x : int32
}

type E = uint16

type Y = {
  e : E
}

type MyFreakingVariant =
| A of A
| X of X
| Y of Y

type C = {
  e : Empty
  c : MyFreakingVariant
}

type Variant_B_A_B_C_Empty_E =
| A of A
| B of B
| C of C
| Empty

type FullTest = {
  primitives : Primitives
  v1 : string array
  v2 : Primitives array
  p : string * Primitives
  o : Primitives option
  q : Variant_B_A_B_C_Empty_E
}
