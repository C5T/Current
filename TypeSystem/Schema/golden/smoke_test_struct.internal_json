{
  "types": [
    [
      "T9000000000000000011",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000011"
        }
      }
    ],
    [
      "T9000000000000000021",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000021"
        }
      }
    ],
    [
      "T9000000000000000022",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000022"
        }
      }
    ],
    [
      "T9000000000000000023",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000023"
        }
      }
    ],
    [
      "T9000000000000000024",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000024"
        }
      }
    ],
    [
      "T9000000000000000031",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000031"
        }
      }
    ],
    [
      "T9000000000000000032",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000032"
        }
      }
    ],
    [
      "T9000000000000000033",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000033"
        }
      }
    ],
    [
      "T9000000000000000034",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000034"
        }
      }
    ],
    [
      "T9000000000000000041",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000041"
        }
      }
    ],
    [
      "T9000000000000000042",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000042"
        }
      }
    ],
    [
      "T9000000000000000051",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000051"
        }
      }
    ],
    [
      "T9000000000000000052",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000052"
        }
      }
    ],
    [
      "T9000000000000000061",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000061"
        }
      }
    ],
    [
      "T9000000000000000062",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000062"
        }
      }
    ],
    [
      "T9010000003568589458",
      {
        "ReflectedType_Enum": {
          "type_id": "T9010000003568589458",
          "name": "E",
          "underlying_type": "T9000000000000000022"
        }
      }
    ],
    [
      "T9200000002835747520",
      {
        "ReflectedType_Struct": {
          "type_id": "T9200000002835747520",
          "native_name": "Empty",
          "fields": []
        }
      }
    ],
    [
      "T9200642690288147741",
      {
        "ReflectedType_Struct": {
          "type_id": "T9200642690288147741",
          "native_name": "FullTest",
          "fields": [
            {
              "type_id": "T9206969065948310524",
              "name": "primitives",
              "description": "A structure with a lot of primitive types."
            },
            {
              "type_id": "T9319767778871345491",
              "name": "v1"
            },
            {
              "type_id": "T9315520306457829347",
              "name": "v2"
            },
            {
              "type_id": "T9334519405000386225",
              "name": "p"
            },
            {
              "type_id": "T9211849004702662543",
              "name": "o"
            },
            {
              "type_id": "T9228482442669086788",
              "name": "q",
              "description": "Field | descriptions | FTW !"
            },
            {
              "type_id": "T9209454265127716773",
              "name": "w1"
            },
            {
              "type_id": "T9209980087718877311",
              "name": "w2"
            },
            {
              "type_id": "T9200915781714511302",
              "name": "w3"
            },
            {
              "type_id": "T9207402181572240291",
              "name": "w4"
            },
            {
              "type_id": "T9209503190895787129",
              "name": "w5"
            },
            {
              "type_id": "T9206651538007828258",
              "name": "w6"
            },
            {
              "type_id": "T9204352959449015213",
              "name": "tsc"
            }
          ]
        }
      }
    ],
    [
      "T9200817599233955266",
      {
        "ReflectedType_Struct": {
          "type_id": "T9200817599233955266",
          "native_name": "B",
          "super_id": "T9206911749438269255",
          "super_name": "A",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "b"
            }
          ]
        }
      }
    ],
    [
      "T9200915781714511302",
      {
        "ReflectedType_Struct": {
          "type_id": "T9200915781714511302",
          "native_name": "Templated_Z",
          "template_inner_id": "T9209626390174323094",
          "template_inner_name": "TemplatedInheriting_Z",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "foo"
            },
            {
              "type_id": "T9209626390174323094",
              "name": "bar"
            }
          ]
        }
      }
    ],
    [
      "T9201673071807149456",
      {
        "ReflectedType_Struct": {
          "type_id": "T9201673071807149456",
          "native_name": "Templated_Z",
          "template_inner_id": "T9200000002835747520",
          "template_inner_name": "Empty",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "foo"
            },
            {
              "type_id": "T9200000002835747520",
              "name": "bar"
            }
          ]
        }
      }
    ],
    [
      "T9202971611369570493",
      {
        "ReflectedType_Struct": {
          "type_id": "T9202971611369570493",
          "native_name": "C",
          "fields": [
            {
              "type_id": "T9200000002835747520",
              "name": "e"
            },
            {
              "type_id": "T9227782344077896555",
              "name": "c"
            },
            {
              "type_id": "T9227782347108675041",
              "name": "d"
            }
          ]
        }
      }
    ],
    [
      "T9204352959449015213",
      {
        "ReflectedType_Struct": {
          "type_id": "T9204352959449015213",
          "native_name": "TrickyEvolutionCases",
          "fields": [
            {
              "type_id": "T9218838894356727119",
              "name": "o1"
            },
            {
              "type_id": "T9218838894356726831",
              "name": "o2"
            },
            {
              "type_id": "T9214663744530229872",
              "name": "o3"
            },
            {
              "type_id": "T9214663744530227568",
              "name": "o4"
            },
            {
              "type_id": "T9214071600727148400",
              "name": "o5"
            },
            {
              "type_id": "T9330011180795395761",
              "name": "o6"
            },
            {
              "type_id": "T9340011180795395761",
              "name": "o7"
            }
          ]
        }
      }
    ],
    [
      "T9206651538007828258",
      {
        "ReflectedType_Struct": {
          "type_id": "T9206651538007828258",
          "native_name": "TemplatedInheriting_Z",
          "super_id": "T9206911749438269255",
          "super_name": "A",
          "template_inner_id": "T9201673071807149456",
          "template_inner_name": "Templated_Z",
          "fields": [
            {
              "type_id": "T9000000000000000042",
              "name": "baz"
            },
            {
              "type_id": "T9201673071807149456",
              "name": "meh"
            }
          ]
        }
      }
    ],
    [
      "T9206911749438269255",
      {
        "ReflectedType_Struct": {
          "type_id": "T9206911749438269255",
          "native_name": "A",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "a"
            }
          ]
        }
      }
    ],
    [
      "T9206969065948310524",
      {
        "ReflectedType_Struct": {
          "type_id": "T9206969065948310524",
          "native_name": "Primitives",
          "fields": [
            {
              "type_id": "T9000000000000000021",
              "name": "a",
              "description": "It's the \"order\" of fields that matters."
            },
            {
              "type_id": "T9000000000000000022",
              "name": "b",
              "description": "Field descriptions can be set in any order."
            },
            {
              "type_id": "T9000000000000000023",
              "name": "c"
            },
            {
              "type_id": "T9000000000000000024",
              "name": "d"
            },
            {
              "type_id": "T9000000000000000031",
              "name": "e"
            },
            {
              "type_id": "T9000000000000000032",
              "name": "f"
            },
            {
              "type_id": "T9000000000000000033",
              "name": "g"
            },
            {
              "type_id": "T9000000000000000034",
              "name": "h"
            },
            {
              "type_id": "T9000000000000000041",
              "name": "i"
            },
            {
              "type_id": "T9000000000000000042",
              "name": "j"
            },
            {
              "type_id": "T9000000000000000051",
              "name": "k"
            },
            {
              "type_id": "T9000000000000000052",
              "name": "l"
            },
            {
              "type_id": "T9000000000000000011",
              "name": "m",
              "description": "Multiline\ndescriptions\ncan be used."
            },
            {
              "type_id": "T9000000000000000061",
              "name": "n"
            },
            {
              "type_id": "T9000000000000000062",
              "name": "o"
            }
          ]
        }
      }
    ],
    [
      "T9207402181572240291",
      {
        "ReflectedType_Struct": {
          "type_id": "T9207402181572240291",
          "native_name": "TemplatedInheriting_Z",
          "super_id": "T9206911749438269255",
          "super_name": "A",
          "template_inner_id": "T9209980946934124423",
          "template_inner_name": "X",
          "fields": [
            {
              "type_id": "T9000000000000000042",
              "name": "baz"
            },
            {
              "type_id": "T9209980946934124423",
              "name": "meh"
            }
          ]
        }
      }
    ],
    [
      "T9208828720332602574",
      {
        "ReflectedType_Struct": {
          "type_id": "T9208828720332602574",
          "native_name": "Y",
          "fields": [
            {
              "type_id": "T9010000003568589458",
              "name": "e"
            }
          ]
        }
      }
    ],
    [
      "T9209454265127716773",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209454265127716773",
          "native_name": "Templated_Z",
          "template_inner_id": "T9209980946934124423",
          "template_inner_name": "X",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "foo"
            },
            {
              "type_id": "T9209980946934124423",
              "name": "bar"
            }
          ]
        }
      }
    ],
    [
      "T9209503190895787129",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209503190895787129",
          "native_name": "TemplatedInheriting_Z",
          "super_id": "T9206911749438269255",
          "super_name": "A",
          "template_inner_id": "T9227782344077896555",
          "template_inner_name": "MyFreakingVariant",
          "fields": [
            {
              "type_id": "T9000000000000000042",
              "name": "baz"
            },
            {
              "type_id": "T9227782344077896555",
              "name": "meh"
            }
          ]
        }
      }
    ],
    [
      "T9209626390174323094",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209626390174323094",
          "native_name": "TemplatedInheriting_Z",
          "super_id": "T9206911749438269255",
          "super_name": "A",
          "template_inner_id": "T9200000002835747520",
          "template_inner_name": "Empty",
          "fields": [
            {
              "type_id": "T9000000000000000042",
              "name": "baz"
            },
            {
              "type_id": "T9200000002835747520",
              "name": "meh"
            }
          ]
        }
      }
    ],
    [
      "T9209827283478105543",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209827283478105543",
          "native_name": "B2",
          "super_id": "T9206911749438269255",
          "super_name": "A",
          "fields": []
        }
      }
    ],
    [
      "T9209980087718877311",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209980087718877311",
          "native_name": "Templated_Z",
          "template_inner_id": "T9227782344077896555",
          "template_inner_name": "MyFreakingVariant",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "foo"
            },
            {
              "type_id": "T9227782344077896555",
              "name": "bar"
            }
          ]
        }
      }
    ],
    [
      "T9209980946934124423",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209980946934124423",
          "native_name": "X",
          "fields": [
            {
              "type_id": "T9000000000000000033",
              "name": "x"
            }
          ]
        }
      }
    ],
    [
      "T9210014876381341935",
      {
        "ReflectedType_Optional": {
          "type_id": "T9210014876381341935",
          "optional_type": "T9206911749438269255"
        }
      }
    ],
    [
      "T9211849004702662543",
      {
        "ReflectedType_Optional": {
          "type_id": "T9211849004702662543",
          "optional_type": "T9206969065948310524"
        }
      }
    ],
    [
      "T9214071600727148400",
      {
        "ReflectedType_Optional": {
          "type_id": "T9214071600727148400",
          "optional_type": "T9315061774377499195"
        }
      }
    ],
    [
      "T9214663744530227568",
      {
        "ReflectedType_Optional": {
          "type_id": "T9214663744530227568",
          "optional_type": "T9319767778871345419"
        }
      }
    ],
    [
      "T9214663744530229872",
      {
        "ReflectedType_Optional": {
          "type_id": "T9214663744530229872",
          "optional_type": "T9319767778871345491"
        }
      }
    ],
    [
      "T9218838894356726831",
      {
        "ReflectedType_Optional": {
          "type_id": "T9218838894356726831",
          "optional_type": "T9000000000000000033"
        }
      }
    ],
    [
      "T9218838894356727119",
      {
        "ReflectedType_Optional": {
          "type_id": "T9218838894356727119",
          "optional_type": "T9000000000000000042"
        }
      }
    ],
    [
      "T9227782344077896555",
      {
        "ReflectedType_Variant": {
          "type_id": "T9227782344077896555",
          "name": "MyFreakingVariant",
          "cases": [
            "T9206911749438269255",
            "T9209980946934124423",
            "T9208828720332602574"
          ]
        }
      }
    ],
    [
      "T9227782347108675041",
      {
        "ReflectedType_Variant": {
          "type_id": "T9227782347108675041",
          "name": "Variant_B_A_X_Y_E",
          "cases": [
            "T9206911749438269255",
            "T9209980946934124423",
            "T9208828720332602574"
          ]
        }
      }
    ],
    [
      "T9228482442669086788",
      {
        "ReflectedType_Variant": {
          "type_id": "T9228482442669086788",
          "name": "Variant_B_A_B_B2_C_Empty_E",
          "cases": [
            "T9206911749438269255",
            "T9200817599233955266",
            "T9209827283478105543",
            "T9202971611369570493",
            "T9200000002835747520"
          ]
        }
      }
    ],
    [
      "T9315061774377499195",
      {
        "ReflectedType_Vector": {
          "type_id": "T9315061774377499195",
          "element_type": "T9206911749438269255"
        }
      }
    ],
    [
      "T9315520306457829347",
      {
        "ReflectedType_Vector": {
          "type_id": "T9315520306457829347",
          "element_type": "T9206969065948310524"
        }
      }
    ],
    [
      "T9319767778871345419",
      {
        "ReflectedType_Vector": {
          "type_id": "T9319767778871345419",
          "element_type": "T9000000000000000033"
        }
      }
    ],
    [
      "T9319767778871345491",
      {
        "ReflectedType_Vector": {
          "type_id": "T9319767778871345491",
          "element_type": "T9000000000000000042"
        }
      }
    ],
    [
      "T9330011180795395761",
      {
        "ReflectedType_Pair": {
          "type_id": "T9330011180795395761",
          "first_type": "T9000000000000000042",
          "second_type": "T9210014876381341935"
        }
      }
    ],
    [
      "T9334519405000386225",
      {
        "ReflectedType_Pair": {
          "type_id": "T9334519405000386225",
          "first_type": "T9000000000000000042",
          "second_type": "T9206969065948310524"
        }
      }
    ],
    [
      "T9340011180795395761",
      {
        "ReflectedType_Map": {
          "type_id": "T9340011180795395761",
          "key_type": "T9000000000000000042",
          "value_type": "T9210014876381341935"
        }
      }
    ]
  ],
  "order": [
    "T9206969065948310524",
    "T9319767778871345491",
    "T9315520306457829347",
    "T9334519405000386225",
    "T9211849004702662543",
    "T9206911749438269255",
    "T9200817599233955266",
    "T9209827283478105543",
    "T9200000002835747520",
    "T9209980946934124423",
    "T9010000003568589458",
    "T9208828720332602574",
    "T9227782344077896555",
    "T9227782347108675041",
    "T9202971611369570493",
    "T9228482442669086788",
    "T9209454265127716773",
    "T9209980087718877311",
    "T9209626390174323094",
    "T9200915781714511302",
    "T9207402181572240291",
    "T9209503190895787129",
    "T9201673071807149456",
    "T9206651538007828258",
    "T9218838894356727119",
    "T9218838894356726831",
    "T9214663744530229872",
    "T9319767778871345419",
    "T9214663744530227568",
    "T9315061774377499195",
    "T9214071600727148400",
    "T9210014876381341935",
    "T9330011180795395761",
    "T9340011180795395761",
    "T9204352959449015213",
    "T9200642690288147741"
  ]
}