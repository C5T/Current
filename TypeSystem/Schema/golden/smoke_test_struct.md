# Data Dictionary

### `Primitives`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (8-bit unsigned) | It's the "order" of fields that matters. |
| `b` | Integer (16-bit unsigned) | Field descriptions can be set in any order. |
| `c` | Integer (32-bit unsigned) |
| `d` | Integer (64-bit unsigned) |
| `e` | Integer (8-bit signed) |
| `f` | Integer (16-bit signed) |
| `g` | Integer (32-bit signed) |
| `h` | Integer (64-bit signed) |
| `i` | Character |
| `j` | String |
| `k` | Number (floating point, single precision) |
| `l` | Number (floating point, double precision) |
| `m` | `true` or `false` | Multiline<br>descriptions<br>can be used. |
| `n` | Time (microseconds since epoch) |
| `o` | Time (milliseconds since epoch) |


### `A`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |


### `B`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |
| `b` | Integer (32-bit signed) |


### `B2`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |


### `X`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `x` | Integer (32-bit signed) |


### `Y`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `e` | Index `E`, underlying type `Integer (16-bit unsigned)` |


### `MyFreakingVariant`
Algebraic type, `A` or `X` or `Y`


### `C`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `e` | `Empty` |
| `c` | Algebraic `A` / `X` / `Y` (a.k.a. `MyFreakingVariant`) |


### `Variant_B_A_B_B2_C_Empty_E`
Algebraic type, `A` or `B` or `B2` or `C` or `Empty`


### `Templated_T9209980946934124423`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `foo` | Integer (32-bit signed) |
| `bar` | `X` |


### `Templated_T9227782344077896555`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `foo` | Integer (32-bit signed) |
| `bar` | Algebraic `A` / `X` / `Y` (a.k.a. `MyFreakingVariant`) |


### `TemplatedInheriting_T9200000002835747520`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |
| `baz` | String |
| `meh` | `Empty` |


### `Templated_T9202973911416238761`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `foo` | Integer (32-bit signed) |
| `bar` | `TemplatedInheriting_T9200000002835747520` |


### `TemplatedInheriting_T9209980946934124423`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |
| `baz` | String |
| `meh` | `X` |


### `TemplatedInheriting_T9227782344077896555`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |
| `baz` | String |
| `meh` | Algebraic `A` / `X` / `Y` (a.k.a. `MyFreakingVariant`) |


### `Templated_T9200000002835747520`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `foo` | Integer (32-bit signed) |
| `bar` | `Empty` |


### `TemplatedInheriting_T9201673071807149456`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `a` | Integer (32-bit signed) |
| `baz` | String |
| `meh` | `Templated_T9200000002835747520` |


### `FullTest`
| **Field** | **Type** | **Description** |
| ---: | :--- | :--- |
| `primitives` | `Primitives` | A structure with a lot of primitive types. |
| `v1` | Array of String |
| `v2` | Array of `Primitives` |
| `p` | Pair of String and `Primitives` |
| `o` | `null` or `Primitives` |
| `q` | Algebraic `A` / `B` / `B2` / `C` / `Empty` (a.k.a. `Variant_B_A_B_B2_C_Empty_E`) | Field &#124; descriptions &#124; FTW ! |
| `w1` | `Templated_T9209980946934124423` |
| `w2` | `Templated_T9227782344077896555` |
| `w3` | `Templated_T9202973911416238761` |
| `w4` | `TemplatedInheriting_T9209980946934124423` |
| `w5` | `TemplatedInheriting_T9227782344077896555` |
| `w6` | `TemplatedInheriting_T9201673071807149456` |

