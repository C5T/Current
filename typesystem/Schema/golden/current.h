// This file is necessary, as an autogenerated schema header file is `#include`-d from the test.
// It could just `#include "../../../current.h"`, but the test `#include`-s the autogenerated schema
// from within a namespace, making things slightly more tricky. -- D.K.

#include <chrono>

#include "../../../typesystem/struct.h"
#include "../../../typesystem/optional.h"
#include "../../../typesystem/variant.h"
#include "../../../typesystem/Evolution/type_evolution.h"
