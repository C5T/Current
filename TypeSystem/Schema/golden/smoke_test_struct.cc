// The `current.h` file is the one from `https://github.com/C5T/Current`.
// Compile with `-std=c++11` or higher.

#ifndef CURRENT_USERSPACE_C9182CD4EAC26CB5
#define CURRENT_USERSPACE_C9182CD4EAC26CB5

#include "current.h"

// clang-format off

namespace current_userspace_c9182cd4eac26cb5 {
struct Primitives {
  // It's the "order" of fields that matters.
  uint8_t a;

  // Field descriptions can be set in any order.
  uint16_t b;
  uint32_t c;
  uint64_t d;
  int8_t e;
  int16_t f;
  int32_t g;
  int64_t h;
  char i;
  std::string j;
  float k;
  double l;

  // Multiline
  // descriptions
  // can be used.
  bool m;
  std::chrono::microseconds n;
  std::chrono::milliseconds o;
};
struct A {
  int32_t a;
};
struct B : A {
  int32_t b;
};
struct B2 : A {
};
struct Empty {
};
struct X {
  int32_t x;
};
enum class E : uint16_t {};
struct Y {
  E e;
};
using MyFreakingVariant = Variant<A, X, Y>;
struct C {
  Empty e;
  MyFreakingVariant c;
};
using Variant_B_A_B_B2_C_Empty_E = Variant<A, B, B2, C, Empty>;
struct Templated_T9209980946934124423 {
  int32_t foo;
  X bar;
};
struct Templated_T9227782344077896555 {
  int32_t foo;
  MyFreakingVariant bar;
};
struct TemplatedInheriting_T9200000002835747520 : A {
  std::string baz;
  Empty meh;
};
struct Templated_T9202973911416238761 {
  int32_t foo;
  TemplatedInheriting_T9200000002835747520 bar;
};
struct TemplatedInheriting_T9209980946934124423 : A {
  std::string baz;
  X meh;
};
struct TemplatedInheriting_T9227782344077896555 : A {
  std::string baz;
  MyFreakingVariant meh;
};
struct Templated_T9200000002835747520 {
  int32_t foo;
  Empty bar;
};
struct TemplatedInheriting_T9201673071807149456 : A {
  std::string baz;
  Templated_T9200000002835747520 meh;
};
struct FullTest {
  // A structure with a lot of primitive types.
  Primitives primitives;
  std::vector<std::string> v1;
  std::vector<Primitives> v2;
  std::pair<std::string, Primitives> p;
  Optional<Primitives> o;

  // Field | descriptions | FTW !
  Variant_B_A_B_B2_C_Empty_E q;
  Templated_T9209980946934124423 w1;
  Templated_T9227782344077896555 w2;
  Templated_T9202973911416238761 w3;
  TemplatedInheriting_T9209980946934124423 w4;
  TemplatedInheriting_T9227782344077896555 w5;
  TemplatedInheriting_T9201673071807149456 w6;
};
}  // namespace current_userspace_c9182cd4eac26cb5

// clang-format on

#endif  // CURRENT_USERSPACE_C9182CD4EAC26CB5
