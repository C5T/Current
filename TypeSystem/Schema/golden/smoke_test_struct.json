{
  "types": [
    [
      "T9000000000000000011",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000011"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000021",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000021"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000022",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000022"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000023",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000023"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000024",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000024"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000031",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000031"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000032",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000032"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000033",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000033"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000034",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000034"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000041",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000041"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000042",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000042"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000051",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000051"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000052",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000052"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000061",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000061"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9000000000000000062",
      {
        "ReflectedType_Primitive": {
          "type_id": "T9000000000000000062"
        },
        "": "T9200000000887757410"
      }
    ],
    [
      "T9200377170767362844",
      {
        "ReflectedType_Struct": {
          "type_id": "T9200377170767362844",
          "name": "FullTest",
          "super_id": "T1",
          "fields": [
            [
              "T9206969065948310524",
              "primitives"
            ],
            [
              "T9319767778871345491",
              "v1"
            ],
            [
              "T9315520306457829347",
              "v2"
            ],
            [
              "T9334519405000386225",
              "p"
            ],
            [
              "T9211849004702662543",
              "o"
            ],
            [
              "T9229638386046851355",
              "q"
            ],
            [
              "T9239638386046851355",
              "r"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9202573821547132669",
      {
        "ReflectedType_Struct": {
          "type_id": "T9202573821547132669",
          "name": "B",
          "super_id": "T9206911749438269255",
          "fields": [
            [
              "T9000000000000000033",
              "b"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9204200194570452885",
      {
        "ReflectedType_Struct": {
          "type_id": "T9204200194570452885",
          "name": "C",
          "super_id": "T1",
          "fields": [
            [
              "T9230545412293788326",
              "c"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9206911749438269255",
      {
        "ReflectedType_Struct": {
          "type_id": "T9206911749438269255",
          "name": "A",
          "super_id": "T1",
          "fields": [
            [
              "T9000000000000000033",
              "a"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9206969065948310524",
      {
        "ReflectedType_Struct": {
          "type_id": "T9206969065948310524",
          "name": "Primitives",
          "super_id": "T1",
          "fields": [
            [
              "T9000000000000000021",
              "a"
            ],
            [
              "T9000000000000000022",
              "b"
            ],
            [
              "T9000000000000000023",
              "c"
            ],
            [
              "T9000000000000000024",
              "d"
            ],
            [
              "T9000000000000000031",
              "e"
            ],
            [
              "T9000000000000000032",
              "f"
            ],
            [
              "T9000000000000000033",
              "g"
            ],
            [
              "T9000000000000000034",
              "h"
            ],
            [
              "T9000000000000000041",
              "i"
            ],
            [
              "T9000000000000000042",
              "j"
            ],
            [
              "T9000000000000000051",
              "k"
            ],
            [
              "T9000000000000000052",
              "l"
            ],
            [
              "T9000000000000000011",
              "m"
            ],
            [
              "T9000000000000000061",
              "n"
            ],
            [
              "T9000000000000000062",
              "o"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9209980946011848465",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209980946011848465",
          "name": "Y",
          "super_id": "T1",
          "fields": [
            [
              "T9000000000000000033",
              "x"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9209980946934124423",
      {
        "ReflectedType_Struct": {
          "type_id": "T9209980946934124423",
          "name": "X",
          "super_id": "T1",
          "fields": [
            [
              "T9000000000000000033",
              "x"
            ]
          ]
        },
        "": "T9204383916472223645"
      }
    ],
    [
      "T9211849004702662543",
      {
        "ReflectedType_Optional": {
          "type_id": "T9211849004702662543",
          "optional_type": "T9206969065948310524"
        },
        "": "T9206609512303532826"
      }
    ],
    [
      "T9229638386046851355",
      {
        "ReflectedType_Polymorphic": {
          "type_id": "T9229638386046851355",
          "cases": [
            "T9206911749438269255",
            "T9202573821547132669",
            "T9204200194570452885"
          ],
          "required": true
        },
        "": "T9202023158513339362"
      }
    ],
    [
      "T9230545412293788326",
      {
        "ReflectedType_Polymorphic": {
          "type_id": "T9230545412293788326",
          "cases": [
            "T9209980946934124423",
            "T9209980946011848465"
          ],
          "required": false
        },
        "": "T9202023158513339362"
      }
    ],
    [
      "T9239638386046851355",
      {
        "ReflectedType_Polymorphic": {
          "type_id": "T9239638386046851355",
          "cases": [
            "T9206911749438269255",
            "T9202573821547132669",
            "T9204200194570452885"
          ],
          "required": false
        },
        "": "T9202023158513339362"
      }
    ],
    [
      "T9315520306457829347",
      {
        "ReflectedType_Vector": {
          "type_id": "T9315520306457829347",
          "element_type": "T9206969065948310524"
        },
        "": "T9209585172575626540"
      }
    ],
    [
      "T9319767778871345491",
      {
        "ReflectedType_Vector": {
          "type_id": "T9319767778871345491",
          "element_type": "T9000000000000000042"
        },
        "": "T9209585172575626540"
      }
    ],
    [
      "T9334519405000386225",
      {
        "ReflectedType_Pair": {
          "type_id": "T9334519405000386225",
          "first_type": "T9000000000000000042",
          "second_type": "T9206969065948310524"
        },
        "": "T9205594056817681433"
      }
    ]
  ],
  "order": [
    "T9206969065948310524",
    "T9319767778871345491",
    "T9315520306457829347",
    "T9334519405000386225",
    "T9211849004702662543",
    "T9206911749438269255",
    "T9202573821547132669",
    "T9209980946934124423",
    "T9209980946011848465",
    "T9230545412293788326",
    "T9204200194570452885",
    "T9229638386046851355",
    "T9239638386046851355",
    "T9200377170767362844"
  ]
}